### ISSUE TITLE
XMB and XLIFF output the wrong source line for a message in an inline template

### ISSUE THREAD

**Issue by Goodwine** [angular/angular#24912]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Performance issue
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

## Current behavior

XMB and XLIFF(1,2) print the wrong source line for a message in an inline template, [example in repo tests](https://github.com/angular/angular/blob/c8ad9657c9423cc2d07d6cc6a230ef28e2b60f3c/packages/compiler-cli/test/extract_i18n_spec.ts#L47):
```html
<msg id="9038505069473852515"><source>src/comp2.ts:2,3</source>
```

1. The compiler reads `@Component` and sets the value for `template` or `templateUrl`.
2. Then it normalizes things like this:
   ```js
   // not normalized
   { "template": "<foo></foo>" }
   { "templateUrl": "foo.html" }
   // normalized
   { "template": "<foo></foo>", templateUrl: "foo.ts" }
   { "templateUrl": "foo.html", template: "<div>bar</div>" }
   ```
3. The compiler sends the value of `template` to an HTML compiler and gets back an AST.
4. The i18n extractor determines the position of the messages given the html AST.

## Expected behavior

The generated code should have been:
```html
<msg id="9038505069473852515"><source>src/comp2.ts:8,9</source>
```
1. The compiler should store a value (e.g. `templateOffsetRows`) pointing to where in the TS file can the template be found.
2. Same behavior
3. Same behavior
4. Same as before, but the row/line numbers are incremented by `templateOffsetRows`.

## Minimal reproduction of the problem with instructions

bazel test //packages/compiler-cli/test:extract_i18n

## What is the motivation / use case for changing the behavior?

This is a follow up for issue #24884, tools that rely on the correctness of the output will be.. incorrect.

This is IMO less important than the other bug because I expect most people to use html templates instead of inline templates in production. There are exceptions like tiny components, but in most cases those tiny components don't really have any messages to be translated. This is ofc on my experience.

## Environment

Commit: c8ad9657

**Comment by Goodwine** [comment]
I plan to send a PR next week on this, hopefully the plan looks good to you.

**Comment by Goodwine** [comment]
After trying several things (the compiler code is really hard to follow), I could not find a feasible way to do this. I managed to get up to this ([here](https://github.com/angular/angular/blob/8e65891985b98cb0766f3f46b2cdf594a17cd10c/packages/compiler/src/aot/compiler.ts#L873)):
```json
"decorators": {
      "__symbolic": "call",
      "expression": {
         "__symbolic": "resolved",
         "symbol": {
            "filePath": "_tmp/2c55867fb740750ba18c7e6a9c4728a9/tmp.48754/node_modules/@angular/core/core.d.ts",
            "name": "Component",
            "members": []
         },
         "line": 3,
         "character": 5,
         "fileName": "_tmp/2c55867fb740750ba18c7e6a9c4728a9/tmp.48754/src/comp2.ts"
      },
      "arguments": [
         {
            "selector": "basic2",
            "template": "<div title=\"translate me\" i18n-title=\"meaning|desc\"></div>\n      <p id=\"welcomeMessage\">\n    <!--i18n-->\n    Welcome<!--/i18n-->\n    </p>"
         }
      ]
   }
]
```
I realized that there is a series of "simplify" calls, and then only some properties relevant to `@Directive` or `@Component` get extracted [here](https://github.com/angular/angular/blob/8e65891985b98cb0766f3f46b2cdf594a17cd10c/packages/compiler/src/aot/static_reflector.ts#L439).

By forcing the returned `Directive` object to have a key for `line`, I was able to get the position and use it, I was almost happy, but then I realized that the position there matches the start of the component, and the position of each "argument" is not available.

I'd imagine it should be possible to add the line for each argument like:
```json
{
  "arguments": [
    {"key": "template", "expression": {
       "line": 6, "character": 12, "symbol": "..."
    }},
  ]
}
```

But I don't know if you are either limited by the TS compiler, or whether you'd even be OK to having this extra complexity there. Do you have any pointers that could help me find what I need?

**Comment by vicb** [comment]
@Goodwine I think we should try to solve it for the IVY engine only because Angular will switch to IVY in a few months. It's probably not worth spending time doing it on the View Engine. WDYT ?

**Comment by petebacondarwin** [comment]
The new message extractor (#32912) based on `$localize` calls should be able to provide the correct information here. I will ensure that it has appropriate behaviour and tests to prove it.

**Comment by petebacondarwin** [comment]
This should be fixed by the new Ivy extractor. But it may fall foul of the same problems as https://github.com/angular/angular/issues/39195. I am going to close this and track the overall problem there.
